DSpace/dspace-angular
# Edit metadata redesign

## References
* Fixes #1722 
* Fixes #1894 

## Description
This PR redesigns the edit-metadata page for items (and adds support for other DSpaceObjects), solving several issues (see references)

## Instructions for Reviewers
Technical changes:
* Added a new module called `DsoSharedModule`, which can be imported wherever an edit metadata page is required for any DSpaceObject (and possible future shared components)
* Within this module, the main component for metadata edit is `DsoEditMetadataComponent` (`ds-themed-dso-edit-metadata` to support themes)
* `DsoEditMetadataForm` is a helper class dynamically storing information about the object's metadata and its changes made to them. It's also responsible for turning all the metadata changes into an array of Operations which can be sent to the REST API to update our object

Functionality changes:
* Metadata values are grouped by their field
* Metadata values are draggable within their field to change their place property
* Adding a new metadata value will automatically add it to the bottom of their respective field
* Metadata fields are sorted alphabetically

Testing guidelines:
* Edit an item and visit the "edit metadata" tab
* Click the add button to start adding a new value
* You should be able to search through all metadata fields using the field input
* Clicking the "confirm" button to the right of the value should add the new value at the bottom of their respective field and remain marked green
* Clicking the "confirm" button whilst an invalid metadata field is filled in, the user should get a warning underneath the input and the value isn't added
* Clicking "undo" (or "delete" on a confirmed value) should simply remove the new value
* Editing an existing metadata value should mark the value orange
* Marking an existing metadata value as "deleted" should mark the value red
* Moving a metadata value within its field should mark the values orange which's "place" properties are changed by the move
* After a move took place, the user should see an "undo" button below the respective metadata field and clicking it should undo any move changes within this field
* Clicking "discard" at the top or bottom of the table after any changes were made should get rid of these changes and revert the metadata table to its original state
* The "reinstate" button should be available after a discard took place and clicking it should bring back the changes that were previously discarded
* Clicking "save" should send a PATCH request to the REST api containing the changes made by the user (note: move changes are calculated dynamically and the operations might not reflect the order the user moved values in, but should have the same result regardless)
* A success notification should display once the object is successfully updated
* Editing a collection's item template should work the exact same as editing an item

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
